statsmodels/doe/latin.py

```diff
@@ -0,0 +1,200 @@
+
+    # Sample scaling from unit hypercube to feature range
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = sample * (max_ - min_) + min_
+
+    return sample
+
+
+def optimal_design(dim, n_sample, bounds=None, start_design=None, niter=1,
+                   force=False):
+    """Optimal design.
+
+    Optimize the design by doing random permutations to lower the centered
+    discrepancy.
+
+    Centered discrepancy based design show better space filling robustness
+    toward 2D and 3D subprojections. Distance based design better space filling
+    but less robust to subprojections.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    start_design : array_like (n_samples, k_vars)
+        Initial design of experiment to optimize.
+
+    Returns
+    -------
+    sample : array_like (n_samples, k_vars)
+        Optimal Latin hypercube Sampling.
+
+    References
+    ----------
+    [1] Damblin et al., "Numerical studies of space filling designs:
+    optimization of Latin Hypercube Samples and subprojection properties",
+    Journal of Simulation, 2013.
+
+    """
+    doe = start_design
+    if doe is None:
+        doe = orthogonal_latin_hypercube(dim, n_sample, bounds)
+
+    def _perturb_doe(x, sample, bounds):
+        """Disturbe the Design of Experiment.
+
+        Parameters
+        ----------
+        x : list of int
+            It is a list of:
+                idx : int
+                    Index value of the components to compute
+        sample : array_like (n_samples, k_vars)
+            Sample to perturb.
+        bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+            Desired range of transformed data. The transformation apply the
+            bounds on the sample and not the theoretical space, unit cube. Thus
+            min and max values of the sample will coincide with the bounds.
+
+        Returns
+        -------
+        discrepancy : float
+            Centered discrepancy.
+
+        """
+        col, row_1, row_2 = map(int, x)
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
+
+        return discrepancy(sample, bounds)
+
+    # Total number of possible design
+    complexity = dim * n_sample ** 2
+
+    if have_basinhopping and ((complexity > 1e6) or force):
+        bounds_optim = ([0, dim - 1], [0, n_sample - 1], [0, n_sample - 1])
+    else:
+        bounds_optim = (slice(0, dim - 1, 1), slice(0, n_sample - 1, 1),
+                        slice(0, n_sample - 1, 1))
```

Comment:
my interpretation: map(int, ...) truncates the float numbers.
if the larges is dim - 1, then the largest value to get truncated to is dim - 2  (one boundary point is mapped to dim-1 but that might never show up). That means the last column and rows are never used in the perturbation
so the upper bounds should be `dim` and `n_sample` IMO

Reply:
partial correction, AFAIU now, brute with finish=None, should return points on the grid, in that case, map(int, optimum) wouldn't do anything given that we already have int, and the last row, column are at dim -1, n_sample -1.

So this will only be a problem if a continuous optimizer is used.

Reply:
I agree for the `int` truncation, I used it incorrectly as `round`.

I thought about using finish in order to switch optimizer from evolutionary to refine around with gradient. But this might be useless.

Reply:
Regarding the return values, even if points are on the grid they are (integer) floating values so numpy cannot use this for indexing. Or you have another way to treat this?

Reply:
You still need to cast to int. but you need to watch out which floats create the ints. i.e. float range needs to be 1 larger to have float values truncated to the highest index if floats are not integer values already as in brute (without continuous `finish`)

Reply:
Very true, so I switched to: `col, row_1, row_2 = np.round(x).astype(int)`. This way, I can left the original bounds as the rounding will be done properly and not to floor.